src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/TokenBucketRateLimiter.cs

```diff
@@ -0,0 +1,319 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+
+namespace System.Threading.RateLimiting
+{
+    /// <summary>
+    /// <see cref="RateLimiter"/> implementation that replenishes tokens periodically instead of via a release mechanism.
+    /// </summary>
+    public sealed class TokenBucketRateLimiter : RateLimiter
+    {
+        private int _tokenCount;
+        private int _queueCount;
+        private uint _lastReplenishmentTick = (uint)Environment.TickCount;
+
+        private readonly Timer? _renewTimer;
+        private readonly TokenBucketRateLimiterOptions _options;
+        private readonly Deque<RequestRegistration> _queue = new Deque<RequestRegistration>();
+
+        // Use the queue as the lock field so we don't need to allocate another object for a lock and have another field in the object
+        private object Lock => _queue;
+
+        private static readonly RateLimitLease SuccessfulLease = new TokenBucketLease(true, null);
+
+        /// <summary>
+        /// Initializes the <see cref="TokenBucketRateLimiter"/>.
+        /// </summary>
+        /// <param name="options">Options to specify the behavior of the <see cref="TokenBucketRateLimiter"/>.</param>
+        public TokenBucketRateLimiter(TokenBucketRateLimiterOptions options)
+        {
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _tokenCount = options.TokenLimit;
+
+            if (_options.AutoReplenishment)
+            {
+                _renewTimer = new Timer(Replenish, this, _options.ReplenishmentPeriod, _options.ReplenishmentPeriod);
```

Comment:
Does this `Timer` get disposed?

Reply:
Nope, it's one of the things I'm planning on bringing up with API review